### PR TITLE
Iss732 refactor

### PR DESCRIPTION
**Pull request by andrea-celentano**

### CODE REVIEW

**Comment by andrea-celentano** [comment]
This will also bring in all the 2019 Data energy corrections into the master.

**Comment by baltzell** [comment]
Looks like we'll have to wait for power to resume at SLAC ...

**Comment by baltzell** [comment]
When running the standard tests (i.e. "mvn install" without the -DskipTests=true) I'm getting HTTP 502 errors on downloading some test files from www.lcsim.org.

**Comment by normangraf** [comment]
Thanks for pointing that out. I think I didn't see that since I have cached versions of the files. I'm guessing it's a side-effect of the power outage. I'll report this.

Norman

________________________________

**Comment by normangraf** [comment]
Should be back up now.

________________________________

**Comment by baltzell** [comment]
Thanks Norman.  I confirm the server is back up, in that the 502 error seems gone.  Currently I'm getting another error:

java.net.SocketTimeoutException: Read timed out

Is this test supposed to work?  Does it work on master?

**Comment by normangraf** [comment]
Let me check.

________________________________

**Comment by baltzell** [comment]
fyi, looks like another dependency host, srs.slac.stanford.edu may be down now

**Comment by normangraf** [comment]
Hmmm, these issues appear to be specific to JLab ifarm nodes. I am getting the following error:

IO Server returned HTTP response code: 503...

**Comment by baltzell** [comment]
I'm working from home, not at JLab

**Comment by normangraf** [comment]
OK, thanks. That’s another piece of the puzzle.

**Comment by normangraf** [comment]
I'm not even able to wget the file when logged in to JLab ifarm:

> wget http://www.lcsim.org/test/hps-java/ap_recon_0000-fullGBL-new.slcio
--2021-01-21 12:47:24--  http://www.lcsim.org/test/hps-java/ap_recon_0000-fullGBL-new.slcio
Resolving www.lcsim.org (www.lcsim.org)... 134.79.129.86
Connecting to www.lcsim.org (www.lcsim.org)|134.79.129.86|:80... connected.
HTTP request sent, awaiting response... 503 Service Unavailable
2021-01-21 12:47:24 ERROR 503: Service Unavailable.

**Review by normangraf — APPROVED** [review]
Builds and tests successfully. Please proceed.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [inline ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:127]: This doesn't seem like a safe way to decide it is the 2019 data.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [inline ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: This seems like an unsafe way to do this that will have to change after we take the 2021 dataset.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java
Comment (cbravo135) [inline ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java:59]: from 2015??

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java
Comment (cbravo135) [inline ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java:61]: Is it a good idea to hard code these? Seems like a prime candidate for a resource file like xml or yaml, or putting them in the db. I get there are a lot of them so the db solution might be a little messy, so the xml/yaml solution would be my preference.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: A suggestion would be for this to be set in some driver config in the steering file.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java
Comment (cbravo135) [inline ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java:18]: Make these a resource somehow rather than hardcode?

**Review by cbravo135 — DISMISSED** [review]

**Review by baltzell — APPROVED** [review]

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:127]: The corrections are not specific to the 2019 data set. They are applicable to the 4.5 GeV running, so this should be OK.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: The 2021 data set will most likely be taken at a lower beam energy, so this should be OK.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java
Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java:61]: We may do this in the fullness of time, but I don't think this should hold up the current release.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java
Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java:18]: We may do this in the fullness of time, but I don't think this should hold up the current release.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:127]: So we expect HPS will never run at any energy above 4.0 GeV that isn't exactly the same energy as 2019? I don't think this is true.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: So we don't expect to take any data beyond 2021?

Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: Yes, we do, but the 2021 data will be at energies less than 4.0 for which we already have corrections. Whether we ever run at energies much above 4.5 is still to be determined. We can come back to this if and when we need to, but I don't think this should hold up the current release.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (normangraf) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:127]: Whether we ever run at energies much above 4.5 is still to be determined. We can come back to this if and when we need to, but I don't think this should hold up the current release.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (andrea-celentano) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: For any new run we will take, we will need to repeat the full set of: calibrations, position corrections, energy corrections. 

The reason is that these depend on the hit threshold, 30 MeV (2019) is very different than 5 MeV (2015 and 2016). Therefore, no matter what we have here now, we will need to repeat the full exercise, understand the detector behavior, validate corrections obtained from MC with real data, ecc... ecc... 

Only after that exercise will be completed, we will commit the new code to hps-java master branch. At that point, we will decide what do to with this "if beamEnergy > 4.0".  We can refactor all the code, or we can simply put another "if" based on the beamEnergy, or we can use the runRange (but what for MC). Since what we have now will work for 2019, that we still need to cook and analyze, I suggest to just move forward.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection2019.java:61]: It doesn't seem like this is at all holding up the release, currently.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java:18]: Code like this sets precedent for it to continue to be developed like this in the future. The organization of this code could be done in a way such that in the future all of these issues of which correction are applied could be simply resolved by our rather sophisticated interface to mysql. Seems like a waste to not leverage the power of that software for these corrections, as it would make future developments and new calibrations for new runs much more straight forward, that could potentially not even need any changes to the code, or at least very minimized changes.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:203]: I agree, and would just like to see the code written in a way that is more forward looking to putting in new additions like this. What I am asking for here is pretty simple and honestly should not take very much time. It honestly could have been done in roughly the amount of time it took both of you to type those responses to my comment.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterPositionCorrection.java:18]: This is currently not holding up the release, as we still have many open PRs that are planned to go into the release are still progressing. We could take a little bit of time and make a more earnest push to refactor this code, as that is the description of the PR.

File: ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java
Comment (cbravo135) [reply ecal-recon/src/main/java/org/hps/recon/ecal/cluster/ClusterCorrectionUtilities.java:127]: It isn't holding up the release. "Whether we ever run at energies much above 4.5 is still to be determined." This is exactly my point, we have not determined either way, so the safer thing to do is assume we will at some point. It has not been determined that we won't. As the code is written, these new corrections are for 2019 data specifically, yet this switch that determines the behavior is not associated to the year the data was taken.

**Comment by cbravo135** [comment]
I just locally checked out this code and merged with master and it fails tests after being merged. Please pull in master and resolve the failed test. I also noticed that the hodo scope readout driver crashes, but it doesn't regisiter as a test failure, so this test should be fixed to actually detect this failure that this code seems to introduce.
```java.lang.NullPointerException
	at org.hps.readout.hodoscope.HodoscopeDigitizationReadoutDriver.getDAQPedestalConditions(HodoscopeDigitizationReadoutDriver.java:155)
	at org.hps.readout.DigitizationReadoutDriver.readHits(DigitizationReadoutDriver.java:472)
	at org.hps.readout.DigitizationReadoutDriver.process(DigitizationReadoutDriver.java:449)
	at org.lcsim.util.Driver.doProcess(Driver.java:261)
	at org.lcsim.util.Driver.processChildren(Driver.java:271)
	at org.lcsim.util.Driver.process(Driver.java:187)
	at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
	at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
	at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
	at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
	at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
	at org.hps.test.it.HodoscopeReadoutTest.testHodoscopeReadout(HodoscopeReadoutTest.java:104)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:255)
	at junit.framework.TestSuite.run(TestSuite.java:250)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:84)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
org.lcsim.job:SEVERE A fatal error occurred when running the job.
java.lang.RuntimeException: java.lang.NullPointerException
	at org.lcsim.util.loop.LCSimLoop.handleClientError(LCSimLoop.java:194)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:834)
	at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
	at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
	at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
	at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
	at org.hps.test.it.HodoscopeReadoutTest.testHodoscopeReadout(HodoscopeReadoutTest.java:104)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:255)
	at junit.framework.TestSuite.run(TestSuite.java:250)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:84)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.lang.NullPointerException
	at org.hps.readout.hodoscope.HodoscopeDigitizationReadoutDriver.getDAQPedestalConditions(HodoscopeDigitizationReadoutDriver.java:155)
	at org.hps.readout.DigitizationReadoutDriver.readHits(DigitizationReadoutDriver.java:472)
	at org.hps.readout.DigitizationReadoutDriver.process(DigitizationReadoutDriver.java:449)
	at org.lcsim.util.Driver.doProcess(Driver.java:261)
	at org.lcsim.util.Driver.processChildren(Driver.java:271)
	at org.lcsim.util.Driver.process(Driver.java:187)
	at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
	... 25 more

maxTimestamp = -1.0
minTimestamp = 9.99999999E8
Driver                                   0 0.00ms
   HodoscopeReadoutTest$HodoAnalDriver   0 0.00ms
      HodoscopeReadoutTest$HodoHitsDrive 0 0.00ms
      HodoscopeReadoutTest$EcalClustersG 0 0.00ms
      HodoscopeReadoutTest$EcalHitsDrive 0 0.00ms
      HodoscopeReadoutTest$HodoscopeRead 0 0.00ms
      HodoscopeReadoutTest$TrackerHitsEC 0 0.00ms
      HodoscopeReadoutTest$ReadoutTimest 0 0.00ms
      HodoscopeReadoutTest$MCParticleDri 0 0.00ms
      *self                              0 0.00ms
   HodoscopePlotsDriver                  0 0.00ms
   AidaSaveDriver                        0 0.00ms
   *self                                 0 0.00ms
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 130.415 sec - in org.hps.test.it.HodoscopeReadoutTest
Running org.hps.test.it.HodoscopeDataConverterTest
```

**Comment by cbravo135** [comment]
I just checked the master branch and org.hps.test.it.HodoscopeReadoutTest also crashes there, so this isn't the source of that. This is however not the test which fails when this branch is merged with master.

**Comment by baltzell** [comment]
After merging with master, all tests pass for me:

https://userweb.jlab.org/~baltzell/.tmp/iss732.log

**Comment by normangraf** [comment]
After merging with master, all tests pass for me as well.

**Pull request merged**
